File: Semester 6/MA 374 (Financial Engg. Lab)/Lab 3/180123062_AB_q3_1.py
# Question 3 Algorithm 1 (inefficient) Lab Assignment 03
# @AB Satyaprakash, 180123062

# imports
from math import exp, sqrt
import numpy as np
import time

# functions ---------------------------------------------------------------------------


def recursion(n, K, S, T, r, sig, M, u, d, p, q, dt):

    if n == M:
        return max(S - K, 0)

    up = recursion(n + 1, K, S * u, T, r, sig, M, u, d, p, q, dt)
    dn = recursion(n + 1, K, S * d, T, r, sig, M, u, d, p, q, dt)
    pc = (exp(-r * dt)) * (p * up + q * dn)
    return pc


# -------------------------------------------------------------------------------------
# Given Initial Values -- and taking K = 100
S0, K, T, r, sig = 100, 100, 1, 0.08, 0.2
# Can't handle values of M beyond 20 ..
Mlist = [5, 10, 20]

for M in Mlist:
    dt = T/M

    u = exp(sig*sqrt(dt)+(r-sig*sig/2)*dt)
    d = exp(-sig*sqrt(dt)+(r-sig*sig/2)*dt)
    p = (exp(r*dt)-d)/(u-d)
    q = 1-p
    loopbackOptionPrice = recursion(0, K, S0, T, r, sig, M, u, d, p, q, dt)
    print('The initial european call option price for M = {} is {}'.format(M, loopbackOptionPrice))

# Test computation time for M = 15: ------------------------------------------------
dt = T/M
u = exp(sig*sqrt(dt)+(r-sig*sig/2)*dt)
d = exp(-sig*sqrt(dt)+(r-sig*sig/2)*dt)
p = (exp(r*dt)-d)/(u-d)
q = 1-p
start = time.time()
timeTempPrice = recursion(0, K, S0, T, r, sig, M, u, d, p, q, dt)
end = time.time()

print('Computational time for M = 15 is {}s'.format(end-start))

# Question 3 algorithm 1 ends -------------------------------------------------------------

File: Semester 6/MA 374 (Financial Engg. Lab)/Lab 3/180123062_AB_q3_1.cpp
// Question 3 Algorithm 1 Lab Assignment 03
// @AB Satyaprakash, 180123062

// include all libraries
#include <bits/stdc++.h>
using namespace std;

// functions
double recursion(double n, double K, double S, double T, double r, double sig, double M, double u, double d, double p, double q, double dt)
{

    // else if n==M and assign to map and return
    if (n == M)
        return max(S - K, 0.0);

    // else do recursion
    double up = recursion(n + 1, K, S * u, T, r, sig, M, u, d, p, q, dt);
    double dn = recursion(n + 1, K, S * d, T, r, sig, M, u, d, p, q, dt);
    double pc = (exp(-r * dt)) * (p * up + q * dn);
    return pc;
}

// code begins in main
int main()
{
    // Given initial values and taking K = 100
    double S0 = 100, K = 100, T = 1, r = 0.08, sig = 0.2;

    vector<double> M(4);
    // Can't handle values of M beyond 25 ..
    M = {5, 10, 25};

    for (int i = 0; i < (int)M.size(); i++)
    {
        // evaluate dt, u, d, p and q as per formulae given in question
        double dt = T / M[i];
        double u = exp(sig * sqrt(dt) + (r - sig * sig / 2) * dt);
        double d = exp(-sig * sqrt(dt) + (r - sig * sig / 2) * dt);
        double p = (exp(r * dt) - d) / (u - d);
        double q = 1 - p;

        double optionPrice = recursion(0, K, S0, T, r, sig, M[i], u, d, p, q, dt);
        cout << fixed << setprecision(15);
        cout << "The initial european call option price for M = " << (int)M[i] << " is " << optionPrice << endl;
    }

    return 0;
}
